[DOC] streamline the Mirror Maker configuration section (#1906)
strimzi/strimzi-kafka-operator: documentation/book/proc-configuring-mirror-maker.adoc

// Module included in the following assemblies:
//
// assembly-deployment-configuration-kafka-mirror-maker.adoc

[id='configuring-kafka-bridge-{context}']
= Configuring Kafka Mirror Maker

Use the properties of the `KafkaMirrorMaker` resource to configure your Kafka Mirror Maker deployment.

> You can configure access control for producer and consumer using TLS or SASL authentication.

Reviewer: ```suggestion
You can configure access control for producers and consumers using TLS or SASL authentication.
```

This procedure shows a configuration that uses TLS encryption and authentication on the consumer side for access control.

For more information on configuring access control for Kafka Mirror Maker, see xref:assembly-access-configuration-kafka-mirror-maker-{context}[Configuring Kafka Mirror Maker access control].

.Prerequisites

* xref:cluster-operator-str[{ProductName} and Kafka is deployed]
* Source and target Kafka clusters are available

.Procedure

. Edit the `spec` properties for the `KafkaMirrorMaker` resource.
+
The properties you can configure are shown in this example configuration:
+
[source,yaml,subs="+quotes,attributes"]
----
apiVersion: {KafkaApiVersion}
kind: KafkaMirrorMaker
metadata:
  name: my-mirror-maker
spec:

>   # ...

Reviewer 2: I think this is pretty much complete example. I would leave this line out.

  replicas: 3 <1>
  consumer:
    bootstrapServers: my-source-cluster-kafka-bootstrap:9092 <2>
    groupId: "my-group" <3>
    numStreams: 2 <4>
    offsetCommitInterval: 120000 <5>

>     tls: <6>

Reviewer 2: Could we add the TLS and authentication sections also to the producr to make it clear that these are configured independently?

Author: Added

>     trustedCertificates:

Reviewer 2: I think this is missaligned:

```suggestion
      trustedCertificates:
```

>       - secretName: my-cluster-cluster-cert

Reviewer 2: Let's use a secret name matching the Strimzi deployed source cluster:

```suggestion
      - secretName: my-source-cluster-cluster-ca-cert
```

        certificate: ca.crt
    authentication: <7>
      type: tls
      certificateAndKey:
        secretName: my-source-secret
        certificate: public.crt
        key: private.key
    config:
      max.poll.records: 100
      receive.buffer.bytes: 32768
  producer:
    bootstrapServers: my-target-cluster-kafka-bootstrap:9092
    abortOnSendFailure: false <8>
    config: <9>
      compression.type: gzip
      batch.size: 8192
  whitelist: "my-topic|other-topic" <10>

>   resources: <11>

Reviewer 2: We should try to use some more realistic request and limit. Maybe request 1CPU and 2Gi and limit 2CPU and 2Gi?

    requests:
      cpu: "8"
      memory: 64Gi
    limits:
      cpu: "12"
      memory: 128Gi
  logging: <12>
    type: inline
    loggers:

>       _logger.name_: "INFO"

Reviewer 2: I think this is confusing to leave here the placeholder (also with respect to the other issue we had on GitHub the other day where I CCed you where we had a user who did not understand that it should be repaced). Mirror Maker seems to have a single logger `mirrormaker.root.logger` so IMHO you should use that directly.

  readinessProbe: <13>
    initialDelaySeconds: 15
    timeoutSeconds: 5
  livenessProbe:
    initialDelaySeconds: 15
    timeoutSeconds: 5
  metrics: <14>
    lowercaseOutputName: true
    rules:
      - pattern: "kafka.server<type=(.+), name=(.+)PerSec\\w*><>Count"
        name: "kafka_server_$1_$2_total"
      - pattern: "kafka.server<type=(.+), name=(.+)PerSec\\w*,
        topic=(.+)><>Count"
        name: "kafka_server_$1_$2_total"
        labels:
          topic: "$3"
  jvmOptions: <15>

>     "-Xmx": "8g"

Reviewer 2: ```suggestion
    "-Xmx": "1g"
```

>     "-Xms": "8g"

Reviewer 2: ```suggestion
    "-Xms": "1g"
```

>   image: my-org/my-image:latest <16>

Reviewer 2: This is kind of very advanced option. Maybe we should leave it out in this example and leave onl the chapter mentioning it in the properties section? I' a bit afraid it might confuse people to think they need to use this.

Author: I've changed to: 

ADVANCED OPTION: Container image configuration, which is recommended only in special situations. 

The sentence links to the Image Configuration section too.

If you prefer we remove, we can.

  template: <17>
      pod:
        affinity:
          podAntiAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
              - labelSelector:
                  matchExpressions:
                    - key: application
                      operator: In
                      values:
                        - postgresql
                        - mongodb
                topologyKey: "kubernetes.io/hostname"
----
+
<1> The number of replica nodes.
<2> Bootstrap servers for consumer and producer.
<3> Group ID for the consumer.
<4> The number of consumer streams.
<5> The offset auto-commit interval in milliseconds.
<6> TLS encryption with key names under which TLS certificates are stored in X.509 format for consumer or producer.
<7> Authentication for consumer or producer, using the TLS mechanism, as shown here, or a SASL-based SCRAM-SHA-512 or PLAIN mechanism.

> <8> Acknowledge (`true`) or ignore (`false`) a send failure for a message.

Reviewer 2: I think if this is set to true, the Mirror Maker will just exit and the container will restart. Acknowledge probably doesn't really fit here.

Author: Rewritten to: If set to `true`, Kafka Mirror Maker will exit and the container will restart following a send failure for a message.

> <9> Kafka configuration options for consumer and producer.

Reviewer 2: The (9) is now only at producer options. Should we add it to the consumer options as well to match the description?

Author: I've moved the number up to the consumer so that it flags the first time we see the config property.

<10> Topics mirrored from source to target Kafka cluster.
<11> Requests for reservation of supported resources, currently `cpu` and `memory`, and limits to specify the maximum resources that can be consumed.
<12> Specified loggers and log levels added directly (`inline`) or indirectly (`external`) through a ConfigMap. A custom ConfigMap must be placed under the `log4j.properties` or `log4j2.properties` key. You can set the log level to INFO, ERROR, WARN, TRACE, DEBUG, FATAL or OFF.
<13> Healthchecks to know when to restart a container (liveness) and when a container can accept traffic (readiness).
<14> Prometheus metrics, which are enabled with configuration for the Prometheus JMX exporter in this example. You can enable metrics without further configuration using `metrics: {}`.
<15> JVM configuration options to optimize performance for the Virtual Machine (VM) running Kafka Mirror Maker.
<16> Container image configuration.

> <17> Pod scheduling to avoid sharing nodes for critical workloads. Here a pod is scheduled based with anti-affinity so the pod is not scheduled on nodes with the same hostname.

Reviewer 2: You list this as pod Scheduling. But this should be listed as templates / customizing deployment and link there to make it clear there are more options than just affinity.

Author: Okay. I'm going to drop the pod scheduling section in the concepts too and just link to the customization section from here.
